### PR TITLE
s2571 - initial commit of DeletedStorageLocationEventSync

### PR DESCRIPTION
**Pull request by jburke-cadc**

### CODE REVIEW

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import org.opencadc.inventory.DeletedStorageLocationEvent;
+import org.opencadc.inventory.InventoryUtil;
+import org.opencadc.tap.TapClient;
+import org.opencadc.tap.TapRowMapper;
+
+public class DeletedStorageLocationEventSync {
+
+    private static final Logger log = Logger.getLogger(DeletedStorageLocationEventSync.class);
+
+    // TODO sufficient to use 3 digit precision ISO_DATE_FORMAT in DateUtil?
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:92]: yes, use the DateUtil constant

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    public static final String ISO_DATE_FORMAT = "yyyy-MM-dd HH:mm:ss.SSSSSS";
+    private static final DateFormat dateFormat = DateUtil.getDateFormat(ISO_DATE_FORMAT, DateUtil.UTC);
+
+    private static final String DELETED_STORAGE_LOCATION_QUERY =
+            "SELECT id, lastModified, metaChecksum FROM inventory.DeletedStorageLocationEvent WHERE lastModified >= %s";
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:97]: The query has to` order by lastModified` in order to incrementally move forward and for the returned iterator to havethe correct ordering

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedArtifactEventSync.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+public class DeletedArtifactEventSync {
+
+    private static final Logger log = Logger.getLogger(DeletedArtifactEventSync.class);
+
+    private static final DateFormat dateFormat = DateUtil.getDateFormat(DateUtil.ISO_DATE_FORMAT, DateUtil.UTC);
+
+    private static final String DELETED_ARTIFACT_QUERY =
+        "SELECT id, lastModified, metaChecksum FROM inventory.DeletedArtifactEvent "
+            + "WHERE lastModified >= %s order by lastModified";
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedArtifactEventSync.java:96]: The query has to take into account that startTime might be null (on first run)

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+public class DeletedStorageLocationEventSync {
+
+    private static final Logger log = Logger.getLogger(DeletedStorageLocationEventSync.class);
+
+    private static final DateFormat dateFormat = DateUtil.getDateFormat(DateUtil.ISO_DATE_FORMAT, DateUtil.UTC);
+
+    private static final String DELETED_STORAGE_LOCATION_QUERY =
+            "SELECT id, lastModified, metaChecksum FROM inventory.DeletedStorageLocationEvent "
+                + "WHERE lastModified >= %s order by lastModified";
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:96]: The query has to take into account that startTime might be null (on first run)

Comment (at88mph) [reply fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java]: The `%s` date comparison needs to be wrapped in single quotes, too.

Comment (jburke-cadc) [reply fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java]: Can you try pulling this branch again. The last pull request should fix this.

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+import org.apache.log4j.Logger;
+import org.opencadc.inventory.DeletedStorageLocationEvent;
+import org.opencadc.inventory.InventoryUtil;
+import org.opencadc.tap.TapClient;
+import org.opencadc.tap.TapRowMapper;
+
+public class DeletedStorageLocationEventSync {
+
+    private static final Logger log = Logger.getLogger(DeletedStorageLocationEventSync.class);
+
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:91]: DateFormat (and Calendar) classes are not thread safe so static instances should never be used (on principle). Just having a member var makes DeletedStorageLocationEventSync non-thread safe, which is fine as long as you know that is it called by one thread only... but I usually create them as locally as possible (here, that would be as a member of the TapRowMapper instance)

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    private static final DateFormat dateFormat = DateUtil.getDateFormat(DateUtil.ISO_DATE_FORMAT, DateUtil.UTC);
+
+    private static final String DELETED_STORAGE_LOCATION_QUERY =
+            "SELECT id, lastModified, metaChecksum FROM inventory.DeletedStorageLocationEvent "
+                + "WHERE lastModified is not null AND lastModified >= '%s' order by lastModified";
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:96]: The lastModified value in the database is never null. The startTime value you are going to use when executing the query is going to be null the first time events are synced from a specific source.

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    private final TapClient<DeletedStorageLocationEvent> tapClient;
+    private final Date startTime;
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:99]: startTime can be null, so it should be public and not final

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+    public DeletedStorageLocationEventSync(TapClient<DeletedStorageLocationEvent> tapClient, Date startTime) {
+        this.tapClient = tapClient;
+        this.startTime = startTime;
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:103]: startTime is optional: not set in constructor

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java (hunk 5 above)
Comment (jburke-cadc) [reply fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:95]: Moved creating the DateFormat into an instance method. If there's no likelihood of the DateFormat method being overridden, could move it directly into the query building.

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java (hunk 7 above)
Comment (jburke-cadc) [reply fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java]: Made startTime public

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java (hunk 8 above)
Comment (jburke-cadc) [reply fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java]: Removed from the constructor.

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java (hunk 6 above)
Comment (jburke-cadc) [reply fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java]: Got that backwards. Fixed the query.

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedArtifactEventSync.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+     *
+     * @return Iterator over the query results.
+     * @throws InterruptedException thread interrupted
+     * @throws IOException failure to write or read the data stream
+     * @throws ResourceNotFoundException remote resource not found
+     * @throws TransientException temporary failure of TAP service: same call could work in future
+     */
+    public ResourceIterator<DeletedArtifactEvent> getEvents()
+        throws InterruptedException, IOException, ResourceNotFoundException, TransientException {
+
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedArtifactEventSync.java:124]: This use of String.format for building a query is more complicated than the problem. Just use a StringBuilder and `append` the parts so it is readable in one place.

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+     *
+     * @return Iterator over the query results.
+     * @throws InterruptedException thread interrupted
+     * @throws IOException failure to write or read the data stream
+     * @throws ResourceNotFoundException remote resource not found
+     * @throws TransientException temporary failure of TAP service: same call could work in future
+     */
+    public ResourceIterator<DeletedStorageLocationEvent> getEvents()
+            throws InterruptedException, IOException, ResourceNotFoundException, TransientException {
+
```
Comment (pdowler) [inline fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:124]: same comment as for DeletedArtifactSync

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedArtifactEventSync.java (hunk 9 above)
Comment (jburke-cadc) [reply fenwick/src/main/java/org/opencadc/fenwick/DeletedArtifactEventSync.java:121]: Updated to use a StringBuilder to create the query.

File: fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java (hunk 10 above)
Comment (jburke-cadc) [reply fenwick/src/main/java/org/opencadc/fenwick/DeletedStorageLocationEventSync.java:121]: As above.

**Pull request merged**